File: backend/app/api/trade.py
"""交易行 / 拍卖 / 面交 HTTP 路由（M7 L2）。"""

from __future__ import annotations

from fastapi import APIRouter, Depends

from app.core.deps import get_current_user, get_trade_service
from app.db.models import User
from app.schemas.common import success
from app.schemas.social_trade import (
    AuctionBidRequest,
    AuctionCreateRequest,
    BazaarDealRequest,
    FaceTradeConfirmRequest,
    FaceTradeInviteRequest,
    FaceTradeLockRequest,
    FaceTradeOfferRequest,
    TradeListingCreateRequest,
)
from app.services.trade_service import TradeService

router = APIRouter(prefix="/trade", tags=["trade"])


@router.get("/listings", response_model=None)
async def trade_list_listings(
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """交易行列表。"""
    return success(await svc.list_listings(current_user))


@router.post("/listings", response_model=None)
async def trade_create_listing(
    body: TradeListingCreateRequest,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """上架。"""
    return success(
        await svc.create_listing(
            current_user,
            mode=body.mode,
            offer_items=[x.model_dump() for x in body.offer_items],
            price_spirit_stones=body.price_spirit_stones,
            ask_items=[x.model_dump() for x in body.ask_items],
        ),
    )


@router.post("/listings/{listing_id}/buy", response_model=None)
async def trade_buy_listing(
    listing_id: int,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """购买。"""
    return success(await svc.buy_listing(current_user, listing_id))


@router.post("/listings/{listing_id}/cancel", response_model=None)
async def trade_cancel_listing(
    listing_id: int,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """撤单。"""
    return success(await svc.cancel_listing(current_user, listing_id))


@router.get("/auctions", response_model=None)
async def trade_list_auctions(
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """拍卖列表。"""
    return success(await svc.list_auctions(current_user))


@router.post("/auctions", response_model=None)
async def trade_create_auction(
    body: AuctionCreateRequest,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """上架拍卖。"""
    return success(
        await svc.create_auction(
            current_user,
            offer_items=[x.model_dump() for x in body.offer_items],
            start_price=body.start_price,
            duration_sec=body.duration_sec,
        ),
    )


@router.post("/auctions/{lot_id}/bid", response_model=None)
async def trade_bid_auction(
    lot_id: int,
    body: AuctionBidRequest,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """出价。"""
    return success(await svc.bid_auction(current_user, lot_id, amount=body.amount))


@router.post("/face", response_model=None)
async def trade_face_invite(
    body: FaceTradeInviteRequest,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """发起交易（pending_invite）。"""
    return success(
        await svc.face_invite(
            current_user,
            peer_character_id=body.peer_character_id,
            peer_name=body.peer_name,
        ),
    )


@router.get("/face/pending", response_model=None)
async def trade_face_pending(
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """收到的待接受交易邀请（大厅邀请列表）。"""
    return success(await svc.face_list_pending(current_user))


@router.get("/face/invite-options", response_model=None)
async def trade_face_invite_options(
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """交易快捷选人：道友 / 道侣 / 同门 / 师徒。"""
    return success(await svc.face_invite_options(current_user))


@router.get("/face/{session_id}", response_model=None)
async def trade_face_get(
    session_id: int,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """交易会话状态。"""
    return success(await svc.face_get(current_user, session_id))


@router.post("/face/{session_id}/accept", response_model=None)
async def trade_face_accept(
    session_id: int,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """受邀方接受交易 → browsing。"""
    return success(await svc.face_accept(current_user, session_id))


@router.post("/face/{session_id}/reject", response_model=None)
async def trade_face_reject(
    session_id: int,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """受邀方拒绝交易。"""
    return success(await svc.face_reject(current_user, session_id))


@router.post("/face/{session_id}/offer", response_model=None)
async def trade_face_offer(
    session_id: int,
    body: FaceTradeOfferRequest,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """更新交易草稿报价（不托管）。"""
    return success(
        await svc.face_set_offer(
            current_user,
            session_id,
            items=[x.model_dump() for x in body.items],
            spirit_stones=body.spirit_stones,
            version=body.version,
            vessel_offer=(
                body.vessel_offer.model_dump() if body.vessel_offer is not None else None
            ),
        ),
    )


@router.post("/face/{session_id}/lock", response_model=None)
async def trade_face_lock(
    session_id: int,
    body: FaceTradeLockRequest,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """锁定己方报价并托管。"""
    return success(
        await svc.face_lock(current_user, session_id, version=body.version),
    )


@router.post("/face/{session_id}/confirm", response_model=None)
async def trade_face_confirm(
    session_id: int,
    body: FaceTradeConfirmRequest,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """确认交易（双方已锁定）。"""
    return success(
        await svc.face_confirm(current_user, session_id, version=body.version),
    )


@router.post("/face/{session_id}/cancel", response_model=None)
async def trade_face_cancel(
    session_id: int,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """取消交易并退回已托管侧。"""
    return success(await svc.face_cancel(current_user, session_id))


@router.get("/bazaar", response_model=None)
async def trade_bazaar_catalog(
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """NPC 坊市货架（固定售价）与可回收摘要。"""
    return success(await svc.list_bazaar(current_user))


@router.post("/bazaar/buy", response_model=None)
async def trade_bazaar_buy(
    body: BazaarDealRequest,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """向坊市购买固定货架道具。"""
    return success(
        await svc.bazaar_buy(
            current_user,
            item_id=body.item_id,
            quantity=body.quantity,
        ),
    )


@router.post("/bazaar/sell", response_model=None)
async def trade_bazaar_sell(
    body: BazaarDealRequest,
    svc: TradeService = Depends(get_trade_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """向坊市出售道具换灵石。"""
    return success(
        await svc.bazaar_sell(
            current_user,
            item_id=body.item_id,
            quantity=body.quantity,
        ),
    )

File: frontend/src/api/trade.ts
/**
 * M7 L2 交易 HTTP API：listings / auctions / face。
 */
import { http } from './http'
import { envelopeFromAxiosError } from './envelope'
import type { ApiResponse } from '../types/api'
import type {
  AuctionListPayload,
  AuctionMutationResult,
  BazaarCatalogPayload,
  BazaarDealResult,
  FaceInviteOptions,
  FaceMutationResult,
  FacePendingPayload,
  ListingListPayload,
  ListingMutationResult,
  TradeItemLine,
} from '../types/trade'

/** GET /trade/listings */
export async function listListings(): Promise<ApiResponse<ListingListPayload>> {
  try {
    const response =
      await http.get<ApiResponse<ListingListPayload>>('/trade/listings')
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<ListingListPayload>(error)
  }
}

/**
 * POST /trade/listings — 上架一口价或易物。
 *
 * @param body - mode / offer_items / price / ask_items
 */
export async function createListing(body: {
  mode?: string
  offer_items: TradeItemLine[]
  price_spirit_stones?: number
  ask_items?: TradeItemLine[]
}): Promise<ApiResponse<ListingMutationResult>> {
  try {
    const response = await http.post<ApiResponse<ListingMutationResult>>(
      '/trade/listings',
      body,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<ListingMutationResult>(error)
  }
}

/**
 * POST /trade/listings/{id}/buy
 *
 * @param listingId - 挂单 id
 */
export async function buyListing(
  listingId: number,
): Promise<ApiResponse<ListingMutationResult>> {
  try {
    const response = await http.post<ApiResponse<ListingMutationResult>>(
      `/trade/listings/${listingId}/buy`,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<ListingMutationResult>(error)
  }
}

/**
 * POST /trade/listings/{id}/cancel
 *
 * @param listingId - 挂单 id
 */
export async function cancelListing(
  listingId: number,
): Promise<ApiResponse<ListingMutationResult>> {
  try {
    const response = await http.post<ApiResponse<ListingMutationResult>>(
      `/trade/listings/${listingId}/cancel`,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<ListingMutationResult>(error)
  }
}

/** GET /trade/auctions */
export async function listAuctions(): Promise<ApiResponse<AuctionListPayload>> {
  try {
    const response =
      await http.get<ApiResponse<AuctionListPayload>>('/trade/auctions')
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<AuctionListPayload>(error)
  }
}

/**
 * POST /trade/auctions — 上架拍品。
 *
 * @param body - offer_items / start_price / duration_sec
 */
export async function createAuction(body: {
  offer_items: TradeItemLine[]
  start_price: number
  duration_sec?: number | null
}): Promise<ApiResponse<AuctionMutationResult>> {
  try {
    const response = await http.post<ApiResponse<AuctionMutationResult>>(
      '/trade/auctions',
      body,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<AuctionMutationResult>(error)
  }
}

/**
 * POST /trade/auctions/{id}/bid
 *
 * @param lotId - 拍品 id
 * @param body - amount 灵石
 */
export async function bidAuction(
  lotId: number,
  body: { amount: number },
): Promise<ApiResponse<AuctionMutationResult>> {
  try {
    const response = await http.post<ApiResponse<AuctionMutationResult>>(
      `/trade/auctions/${lotId}/bid`,
      body,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<AuctionMutationResult>(error)
  }
}

/**
 * POST /trade/face — 发起交易。
 *
 * @param body - peer_name 与 peer_character_id 二选一
 */
export async function faceInvite(body: {
  peer_name?: string | null
  peer_character_id?: number | null
}): Promise<ApiResponse<FaceMutationResult>> {
  try {
    const response = await http.post<ApiResponse<FaceMutationResult>>(
      '/trade/face',
      body,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<FaceMutationResult>(error)
  }
}

/**
 * GET /trade/face/pending — 收到的待接受交易邀请。
 */
export async function faceListPending(): Promise<ApiResponse<FacePendingPayload>> {
  try {
    const response = await http.get<ApiResponse<FacePendingPayload>>(
      '/trade/face/pending',
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError(error)
  }
}

/**
 * GET /trade/face/invite-options — 快捷选人。
 */
export async function faceInviteOptions(): Promise<ApiResponse<FaceInviteOptions>> {
  try {
    const response = await http.get<ApiResponse<FaceInviteOptions>>(
      '/trade/face/invite-options',
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError(error)
  }
}

/**
 * GET /trade/face/{id}
 *
 * @param sessionId - 交易会话 id
 */
export async function faceGet(
  sessionId: number,
): Promise<ApiResponse<FaceMutationResult>> {
  try {
    const response = await http.get<ApiResponse<FaceMutationResult>>(
      `/trade/face/${sessionId}`,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<FaceMutationResult>(error)
  }
}

/**
 * POST /trade/face/{id}/offer — 更新己方草稿报价（不托管）。
 *
 * @param sessionId - 会话 id
 * @param body - items / spirit_stones / version
 */
export async function faceOffer(
  sessionId: number,
  body: {
    items: TradeItemLine[]
    spirit_stones: number
    vessel_offer?: { hours: number } | null
    version: number
  },
): Promise<ApiResponse<FaceMutationResult>> {
  try {
    const response = await http.post<ApiResponse<FaceMutationResult>>(
      `/trade/face/${sessionId}/offer`,
      body,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<FaceMutationResult>(error)
  }
}

/**
 * POST /trade/face/{id}/accept — 受邀方接受。
 *
 * @param sessionId - 会话 id
 */
export async function faceAccept(
  sessionId: number,
): Promise<ApiResponse<FaceMutationResult>> {
  try {
    const response = await http.post<ApiResponse<FaceMutationResult>>(
      `/trade/face/${sessionId}/accept`,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<FaceMutationResult>(error)
  }
}

/**
 * POST /trade/face/{id}/reject — 受邀方拒绝。
 *
 * @param sessionId - 会话 id
 */
export async function faceReject(
  sessionId: number,
): Promise<ApiResponse<FaceMutationResult>> {
  try {
    const response = await http.post<ApiResponse<FaceMutationResult>>(
      `/trade/face/${sessionId}/reject`,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<FaceMutationResult>(error)
  }
}

/**
 * POST /trade/face/{id}/lock — 锁定并托管己方报价。
 *
 * @param sessionId - 会话 id
 * @param body - version
 */
export async function faceLock(
  sessionId: number,
  body: { version: number },
): Promise<ApiResponse<FaceMutationResult>> {
  try {
    const response = await http.post<ApiResponse<FaceMutationResult>>(
      `/trade/face/${sessionId}/lock`,
      body,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<FaceMutationResult>(error)
  }
}

/**
 * POST /trade/face/{id}/confirm
 *
 * @param sessionId - 会话 id
 * @param body - version
 */
export async function faceConfirm(
  sessionId: number,
  body: { version: number },
): Promise<ApiResponse<FaceMutationResult>> {
  try {
    const response = await http.post<ApiResponse<FaceMutationResult>>(
      `/trade/face/${sessionId}/confirm`,
      body,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<FaceMutationResult>(error)
  }
}

/**
 * POST /trade/face/{id}/cancel
 *
 * @param sessionId - 会话 id
 */
export async function faceCancel(
  sessionId: number,
): Promise<ApiResponse<FaceMutationResult>> {
  try {
    const response = await http.post<ApiResponse<FaceMutationResult>>(
      `/trade/face/${sessionId}/cancel`,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<FaceMutationResult>(error)
  }
}

/** GET /trade/bazaar */
export async function fetchBazaar(): Promise<ApiResponse<BazaarCatalogPayload>> {
  try {
    const response = await http.get<ApiResponse<BazaarCatalogPayload>>('/trade/bazaar')
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<BazaarCatalogPayload>(error)
  }
}

/**
 * POST /trade/bazaar/buy
 *
 * @param itemId - 物品 id
 * @param quantity - 数量
 */
export async function buyFromBazaar(
  itemId: string,
  quantity: number,
): Promise<ApiResponse<BazaarDealResult>> {
  try {
    const response = await http.post<ApiResponse<BazaarDealResult>>(
      '/trade/bazaar/buy',
      { item_id: itemId, quantity },
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<BazaarDealResult>(error)
  }
}

/**
 * POST /trade/bazaar/sell
 *
 * @param itemId - 物品 id
 * @param quantity - 数量
 */
export async function sellToBazaar(
  itemId: string,
  quantity: number,
): Promise<ApiResponse<BazaarDealResult>> {
  try {
    const response = await http.post<ApiResponse<BazaarDealResult>>(
      '/trade/bazaar/sell',
      { item_id: itemId, quantity },
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<BazaarDealResult>(error)
  }
}
